sdb/commands/stacks.py

```diff
@@ -321,56 +387,33 @@ def _call(self, objs: Iterable[drgn.Object]) -> Iterable[drgn.Object]:
+        stack_aggr: Dict[Any, List[drgn.Object]] = Stacks.aggregate_stacks(objs)
         for stack_key, tasks in sorted(stack_aggr.items(),
                                        key=lambda x: len(x[1]),
                                        reverse=True):
-            task_state = stack_key[0]
-            if self.args.tstate and self.args.tstate != task_state:
-                continue
-
-            stacktrace_info = ""
-            if self.args.all:
+            if self.match_stack(stack_key, tasks, print_cb):
                 for task in tasks:
-                    stacktrace_info += "{:<18s} {:<16s}\n".format(
-                        hex(task.value_()), task_state)
-            else:
-                stacktrace_info += "{:<18s} {:<16s} {:6d}\n".format(
-                    hex(tasks[0].value_()), task_state, len(tasks))
-
-            mod_match, func_match = False, False
-
-            #
-            # Note on the type-check being ignored:
-            # The original `stack_key` type is a list where the first
-            # element is a string and the rest of them are integers
-            # but this is not easily expressed in mypy, thus we ignore
-            # the assignment error below.
-            #
-            frame_pcs: List[int] = stack_key[1:]  #type: ignore[assignment]
-            for frame_pc in frame_pcs:
-                if mod_start != -1 and mod_start <= frame_pc < mod_end:
-                    mod_match = True
-
-                try:
-                    sym = sdb.get_symbol(frame_pc)
-                    func, offset = sym.name, frame_pc - sym.address
-                    if self.args.function and self.args.function == func:
-                        func_match = True
-                except LookupError:
-                    func, offset = hex(frame_pc), 0x0
+                    yield task
 
-                #
-                # As a potential future item, we may want to print
-                # the frame with the module where the pc/function
-                # belongs to. For example:
-                #     txg_sync_thread+0x15e [zfs]
-                #
-                stacktrace_info += "{:18s}{}+{}\n".format("", func, hex(offset))
-
-            if mod_start != -1 and not mod_match:
-                continue
-            if self.args.function and not func_match:
-                continue
-            print(stacktrace_info)
-        return []
+    def pretty_print(self, objs: Iterable[drgn.Object]) -> None:
+        self.validate_context()
+        self.print_header()
+        #
+        # gather_threads is a generator, but since pretty_print is not, we
+        # convert the output to a list to drain it.
+        #
+        list(self.gather_threads(objs, print))
+
+    def no_input(self) -> Iterable[drgn.Object]:
+        self.validate_context()
+        yield from self.gather_threads(for_each_task(sdb.get_prog()),
+                                       lambda x: None)
```

Comment:
Wouldn't it be better if we made the declaration of `gather_threads()` have `print_cb` to be `None` as the default/optional argument, this way you wouldn't have to pass the lambda here.

This may give errors but I was thinking something like this:
```
    def gather_threads(
            self, objs: Iterable[drgn.Object],
            print_cb: Optional[Callable[[str], None]] = None) -> Iterable[drgn.Object]:
....
    if print_cb:
        print_cb(stacktrace_info)
    return True
```

Then `no_input` should work like this:
```
yield from self.gather_threads(for_each_task(sdb.get_prog()))
```

I think it is fine leaving it as is too, I just though that the lambda was a bit awkward.

Reply:
Despite our conversation the other day, I don't really understand the benefit of passing the callback, vs. passing a boolean, especially if gather_threads() has to check the argument in either case.

All that said, it would be really nice if we could redesign this such that we can separate gathering from printing (see my other comments on how this could be made to perform well).

Reply:
I like that idea, thanks for the suggestion!